### ISSUE TITLE
OTP to Meet web accessibility standard

### ISSUE THREAD

**Issue by robertgregg3** [opentripplanner/otp-react-redux#343]
# Preface 

The first question we would need to find out is what level of accessibility we need to meet (if that is possible). You can have three levels A, AA and AAA. It then I suppose depends on how they would measure this. If they run the site through an accessibility checker we would need to make sure that all points are satisfied rather than just doing what we can with a list. 

# General Checklist 

## Tabbed navigation

- Make sure that everything that is interactive you can "tab" into 
- There is a visible focus ring around each interactive element 
- OFF Screen interactive content (like when you resize a window and you may be tabbing through a menu but not even know it) should be disabled and removed from the tab order 

 
### Navigate the screen using a screen reader 
- The screen reader should announce the controls that they land on correctly 
- Image Alt text attributes exist for all images 
- Buttons are interactive 
- Navigating through a drop down – use up and down and then press enter to see that the response is from the screen reader is correct.  
- That the screen reader jumps to what you want it to focus on. For example when hitting "plan route" where would we like the focus to go? 

### Page Structure 

- Appropriate headings 
- Appropriate landmark roles and elements on the page 
- Open the router in voice over (web elements) to give you an outline of the page - CTR ALT U to open or CMD + OPTION + U on mac. 
- Check the landmarks to make sure they are structured correctly. 

### Colour and contrast  

- Axe extension - https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd 
- Ends up in the dev tools panel as "Accessibility" menu. Press "analyse" to check your site 
- Chrome accessibility dev tools - https://chrome.google.com/webstore/detail/accessibility-developer-t/fpkknkljclfencbdbgkenhalefipecmb/related?hl=en 
- Similar to the above but offers suggestions for colours 
- Ends up in dev tools with "Accessibility properties" menu 
- Click the swatches to see the colour recommendation difference 

### Regression testing  
- using the axe-core library:  https://github.com/dequelabs/axe-core 

# Action Items RE OTP: 

 
### Tabbed navigation 

- Tab navigation is missing in some of the items when you click into the route options. This could be OFF SCREEN items that we can't see because it takes a few tab presses to get to the next item. 
- It is hard to see some of the tabbed navigation when the advanced options is expanded. Perhaps it would be better for it to move the element below (slide) so that all route options can be seen and not hidden by the advanced options.  
- When a route is on the map, the tab navigation works with the start and end points but not the points in between 
- When keying up and down in a drop down box the narrator says the correct item but the drop down box doesn’t move.  It does move when the screen reader is off. This is on Chrome on my Windows Laptop.  

### Navigate the screen using a screen reader 

- Screen reader says the word "skin" quite a few times.  On an occasion such as "Button, skin". Not sure what that is and if we should correct it or not.  
- There were a few questionable items from the screen reader. Would be good to look at each intractable element and make sure it is returning what we require.  

### Page Structure 

- Will need to look and study each item within the page structure and Identify each item 
and decide on the desired outcomes for each 

### Colour and contrast  

- When running the AXE dev tools there are quite a few items to have a look at: 

 

![data tools axe report](https://user-images.githubusercontent.com/4264701/111642432-ff9abb80-87f5-11eb-8d4b-918783ec7a78.png)

Links for Windows Screen Reader: https://support.microsoft.com/en-us/windows/appendix-b-narrator-keyboard-commands-and-touch-gestures-8bdab3f4-b3e9-4554-7f28-8b15bd37410a 

https://support.microsoft.com/en-us/windows/chapter-5-navigation-68941680-3245-6ef5-5012-0674b8b6fc59
